packages/scheduler/lib/workers/worker.ts

```diff
@@ -0,0 +1,104 @@
+import * as fs from 'fs';
+import type { MessagePort } from 'node:worker_threads';
+import { Worker, isMainThread } from 'node:worker_threads';
+import { stringifyError } from '@nangohq/utils';
+import { logger } from '../utils/logger.js';
+import { setTimeout } from 'node:timers/promises';
+import type knex from 'knex';
+
+interface ExpiredTasksMessage {
+    ids: string[];
+}
+
+export abstract class SchedulerWorker {
+    protected workerUrl: URL;
+    protected name: string;
+    private worker: Worker | null = null;
+    constructor({ name, workerUrl, databaseUrl, databaseSchema }: { name: string; workerUrl: URL; databaseUrl: string; databaseSchema: string }) {
+        if (isMainThread) {
+            this.name = name;
+            this.workerUrl = workerUrl;
+            if (!fs.existsSync(workerUrl)) {
+                throw new Error(`Code for ${name} worker not found at ${workerUrl.href}`);
+            }
+
+            const createWorker = () => {
+                this.worker = new Worker(workerUrl, { workerData: { url: databaseUrl, schema: databaseSchema } });
+
+                this.worker.on('error', (err) => {
+                    logger.error(`${name} worker error: ${stringifyError(err)}`);
+                });
+
+                this.worker.on('exit', (code) => {
+                    logger.error(`${name} worker exited with code: ${code}. Restarting...`);
+                    this.worker = null;
+                    createWorker();
+                    this.start();
+                });
+            };
+            createWorker();
```

Comment:
the only change of logic is here. The rest is just consolidating duplicated code into one place